applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h

```diff
@@ -0,0 +1,435 @@
+            "Invalid number of eigenvalues provided" << std::endl;
+
+        KRATOS_ERROR_IF( mParam["subspace_size"].GetInt() < 0 ) <<
+            "Invalid subspace size provided" << std::endl;
+
+        KRATOS_ERROR_IF( mParam["max_iteration"].GetInt() < 1 ) <<
+            "Invalid maximal number of iterations provided" << std::endl;
+
+        KRATOS_INFO_IF( "FEASTEigensystemSolver",
+            mParam["number_of_eigenvalues"].GetInt() > 0  && mParam["subspace_size"].GetInt() > 0 ) <<
+            "Manually defined subspace size will be overwritten to match the defined number of eigenvalues" << std::endl;
+
+        const TScalarOut T = {};
+        CheckParameters(T);
+    }
+
+    ~FEASTEigensystemSolver() override {}
+
+    /**
+     * Solve the generalized eigenvalue problem using FEAST
+     * @param rK first input matrix
+     * @param rM second input matrix
+     * @param rEigenvalues eigenvalues
+     * @param rEigenvectors row-aligned eigenvectors [n_evs,n_dofs]
+     */
+    void Solve(
+        SparseMatrixType& rK,
+        SparseMatrixType& rM,
+        VectorType& rEigenvalues,
+        DenseMatrixType& rEigenvectors) override
+    {
+        // settings
+        const size_t system_size = rK.size1();
+        size_t subspace_size;
+        const TScalarIn Ti = {};
+        const TScalarOut T = {};
+
+        if( mParam["search_lowest_eigenvalues"].GetBool() || mParam["search_highest_eigenvalues"].GetBool() )
+        {
+            subspace_size = 2 * static_cast<size_t>(mParam["number_of_eigenvalues"].GetInt());
+        }
+        else if( mParam["subspace_size"].GetInt() == 0 )
+        {
+            subspace_size = 1.5 * static_cast<size_t>(mParam["number_of_eigenvalues"].GetInt());
+        }
+        else
+        {
+            subspace_size = static_cast<size_t>(mParam["subspace_size"].GetInt());
+        }
+
+        // create column based matrix for the fortran routine
+        FEASTMatrixType tmp_eigenvectors(system_size, subspace_size);
+        OutputVectorType tmp_eigenvalues(subspace_size);
+        VectorType residual(subspace_size);
+
+        // set FEAST settings
+        int fpm[64] = {};
+        feastinit(fpm);
+
+        mParam["echo_level"].GetInt() > 0 ? fpm[0] = 1 : fpm[0] = 0;
+
+        fpm[2] = -std::log10(mParam["tolerance"].GetDouble());
+        fpm[3] = mParam["max_iteration"].GetInt();
+
+        // compute only right eigenvectors
+        if( !TSymmetric )
+            fpm[14] = 1;
+
+        if( mParam["search_lowest_eigenvalues"].GetBool() )
+        {
+            fpm[39] = -1;
+        }
+        if( mParam["search_highest_eigenvalues"].GetBool() )
+        {
+            fpm[39] = 1;
+        }
+
+        char UPLO = 'F';
+        int N = static_cast<int>(system_size);
+
+        // provide matrices in array form. fortran indices start with 1, must be int
+        double* A = reinterpret_cast<double*>(rK.value_data().begin());
```

Comment:
why do you need the `reinterpret_cast`? shoudln't this already be `double`?

Reply:
no, this can also be complex (for the two complex eigenvalues solvers)